core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -154,4 +299,86 @@ public void copyObjectToServer(InetSocketAddress host, KernelOID oid, KernelObje
+     * them in {@link amino.run.kernel.client.KernelClient#metrics}. These collected metrics are
+     * reported to OMS in heartBeats
+     */
+    private void measureMetrics(final KernelServerInfo serverInfo) {
+        KernelServer server = serverInfo.remoteRef;
+        int dataLength = serverInfo.data.len;
+        NodeMetric metric = serverInfo.metric;
+        int period = serverInfo.metricPollPeriod;
+        serverInfo.refCount++;
+        if (serverInfo.refCount == Integer.MAX_VALUE) {
+            /* Reset to 1 when it is about to overflow */
+            serverInfo.refCountOnHeartBeat = serverInfo.refCount = 1;
+        }
+
+        try {
+            /* Forward time = Request Serialization + Network Delay + Deserialization + Object/call Dispatch time
+            + processing time + Optionally connection establishment time */
+            /* Return time = Response Serialization + Network Delay + Deserialization + Get current nano time at end */
+            /* Forward time would be comparatively higher than return time. But to avoid the clock sync issues
+            between client and server, took the round trip time(RTT) as latency
+             */
+
+            /* Make an empty heartbeat to ensure session is established and cached before measurement */
+            server.receiveHeartBeat();
+
+            /* Measure latency */
+            long t1 = System.nanoTime();
+            server.receiveHeartBeat();
+            long t2 = System.nanoTime();
+
+            /* Measure data transfer rate by sending some arbitrary data */
+            server.receiveHeartBeat(serverInfo.data);
+            long t3 = System.nanoTime();
+            if (t3 - t2 < t2 - t1) {
+                period = MIN_METRIC_POLL_PERIOD_MS;
+                serverInfo.data.len += STEP_SIZE_IN_BYTES;
+                serverInfo.stableDataRateTimes = 0;
+                serverInfo.metricPollPeriod = period;
+                serverInfo.metricsTimer.reset(period);
+                logger.fine(String.format("Data size increased to : %d", serverInfo.data.len));
+                // Ignore this sample
+                return;
+            } else if (serverInfo.stableDataRateTimes
+                    >= KernelServerInfo.MIN_STABLE_DATA_RATE_TIMES) {
+                /* Data rates are consistent. Double the poll period */
+                period = period << 1;
+                /* Limit max poll period to 512 sec */
+                period = period > MAX_METRIC_POLL_PERIOD_MS ? MAX_METRIC_POLL_PERIOD_MS : period;
+                serverInfo.stableDataRateTimes = 0;
+                /* TODO: Can try reducing the length by step size and get to an optimum length required to send in
+                heartbeats. But, need to ensure it do not toggle between the 2 consecutive values when optimal value is
+                reached */
+            }
+
+            serverInfo.stableDataRateTimes++;
+            metric.latency = (t2 - t1); // RTT
```

Comment:
Surely you should set this irrespective of whether you ignored the sample was ignored above.  The measured latency is always correct. The measured throughput is only correct in some cases.

Reply:
Actually node metric object has latency and data rate values in it. So ignoring this latency when data rate cannot be calculated. Otherwise, we need to send node metric object(sample) to OMS with valid latency and data rate value being 0. 
Since, this case can happen for few initial times until data size is significant enough to get data transfer time > latency, ignoring the sample.